### PR TITLE
Refactor simulation initialisation

### PR DESCRIPTION
**Pull request by AstroKriel**
Boilerplate initialization code has been moved to the AMRSimulation class. Two functions are exposed to the user, first, which allows the user to perform a computation before initialization (for example to read in a table), and a second is called while the initialization code iterates over the simulation domain. These functions have empty implementations by default, so the user will need to implement them when relevant.

### CODE REVIEW

File: src/grid.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include <AMReX.H>
```
Comment (BenWibking) [inline src/grid.hpp:4]: This needs an include guard.

File: src/grid.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include <AMReX_Array.H>
+#include <AMReX_Array4.H>
+#include <AMReX_Box.H>
+#include <AMReX_REAL.H>
+
+enum class centering { cc=0, fc, ec };
```
Comment (BenWibking) [inline src/grid.hpp]: put the enums and the struct inside a `namespace quokka { .. }` block

File: src/simulation.hpp [hunk 3]
```diff
@@ -942,6 +969,9 @@ void AMRSimulation<problem_t>::MakeNewLevelFromScratch(
   // set state_new_[lev] to desired initial condition
   setInitialConditionsAtLevel(level);
 
+  amrex::IntVect iv(AMREX_D_DECL(19, 0, 5));
+  print_state(state_new_[level], iv);
+
```
Comment (BenWibking) [inline src/simulation.hpp]: Delete this debugging code.

**Review by BenWibking — CHANGES_REQUESTED** [review]
Looks good overall. The main thing is I think we want to put the `grid` struct inside the `quokka` namespace. This means it will be referenced with `quokka::grid` instead of just `grid`.

File: src/simulation.hpp (hunk 3 above)
Comment (AstroKriel) [reply src/simulation.hpp]: Ugh, I thought I had already deleted this. Done now (again).

File: src/grid.hpp (hunk 2 above)
Comment (AstroKriel) [reply src/grid.hpp]: Done.

File: src/grid.hpp (hunk 1 above)
Comment (AstroKriel) [reply src/grid.hpp:4]: Done.

**Review by BenWibking — APPROVED** [review]

**Comment by BenWibking** [comment]
@AstroKriel can you take a look at the build errors?

**Comment by AstroKriel** [comment]
> @AstroKriel can you take a look at the build errors?

Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.

**Comment by BenWibking** [comment]
> > @AstroKriel can you take a look at the build errors?
> 
> Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.

The GPU build is still failing. Can you take a look at that?

**Comment by AstroKriel** [comment]
> > > @AstroKriel can you take a look at the build errors?
> > 
> > 
> > Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.
> 
> The GPU build is still failing. Can you take a look at that?

It looks like the build error occurred because the reference to the state array was stored in grid_vec[...].array, and that vector lives on the CPU. There isn't an operator implemented to access vector elements stored on the CPU from the GPU, and that is why the error was being thrown. I fixed this by extracting the reference to the state variable and passing that reference directly to computeWaveSolution(...) instead of accessing the reference via the grid_vec[...] vector in the function call.

**Comment by BenWibking** [comment]
> It looks like the build error occurred because the reference to the state array was stored in grid_vec[...].array, and that vector lives on the CPU. There isn't an operator implemented to access vector elements stored on the CPU from the GPU, and that is why the error was being thrown. I fixed this by extracting the reference to the state variable and passing that reference directly to computeWaveSolution(...) instead of accessing the reference via the grid_vec[...] vector in the function call.

Yeah, the fact that `std::vector` is not usable from the GPU requires a lot of workarounds in general. This looks like a reasonable way to do it. I'll re-run the GPU tests and merge if it passes.

**Comment by BenWibking** [comment]
@AstroKriel something is still not right. GPU build still fails.

**Comment by AstroKriel** [comment]
Ah yep, it looks like the state reference (ie grid_vec[...].array) needs to be extracted outside all the ParFor loops, because when compiled on GPU, the same problem as before occurs. I'll fix this and see if the problem persists / if there are other problems.

**Comment by BenWibking** [comment]
@AstroKriel there are GPU build failures in the HydroBlast3D problem.

**Pull request merged**